extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -0,0 +1,222 @@
+    final List<VirtualColumn> frameWriterVirtualColumns = new ArrayList<>();
+    frameWriterVirtualColumns.add(partitionBoostVirtualColumn);
+
+    final VirtualColumn segmentGranularityVirtualColumn =
+        QueryKitUtils.makeSegmentGranularityVirtualColumn(jsonMapper, query);
+
+    if (segmentGranularityVirtualColumn != null) {
+      frameWriterVirtualColumns.add(segmentGranularityVirtualColumn);
+    }
+
+    this.frameWriterVirtualColumns = VirtualColumns.create(frameWriterVirtualColumns);
+  }
+
+  // deep storage
+  @Override
+  protected ReturnOrAwait<Unit> runWithSegment(SegmentWithDescriptor segment) throws IOException
+  {
+    return null;
+  }
+
+  // realtime
+  @Override
+  protected ReturnOrAwait<Unit> runWithLoadedSegment(SegmentWithDescriptor segment) throws IOException
+  {
+    return null;
+  }
+
+  // previous stage output
+  @Override
+  protected ReturnOrAwait<Unit> runWithInputChannel(ReadableFrameChannel inputChannel, FrameReader inputFrameReader)
+      throws IOException
+  {
+    // Read the frames from the channel
+    // convert to FrameRowsAndColumns
+
+    if (inputChannel.canRead()) {
+      Frame f = inputChannel.read();
+
+      // the frame here is row based
+      // frame rows and columns need columnar. discuss with Eric
+      // Action item: need to implement a new rows and columns that accept a row-based frame
+
+      // Create a frame rows and columns what would
+      RowBasedFrameRowAndColumns frameRowsAndColumns = new RowBasedFrameRowAndColumns(f, inputFrameReader.signature());
+      Operator op = getOperator(frameRowsAndColumns);
+      //
+      //Operator op = new SegmentToRowsAndColumnsOperator(frameSegment);
+      // On the operator created above add the operators present in the query that we want to chain
+
+      for (OperatorFactory of : query.getOperators()) {
+        op = of.wrap(op);
+      }
+
+      // Let the operator run
+      // the results that come in the receiver must be pushed to the outout channel
+      // need to transform the output rows and columns back to frame
+      Operator.go(op, new Operator.Receiver()
+      {
+        @Override
+        public Operator.Signal push(RowsAndColumns rac)
+        {
+          //outputFrameChannel.output(rac.toFrame());
+          LazilyDecoratedRowsAndColumns tmpldrc = new LazilyDecoratedRowsAndColumns(
+              rac,
+              null,
+              null,
+              null,
+              OffsetLimit.limit(Integer.MAX_VALUE),
+              null,
+              null
+          );
+          Pair<byte[], RowSignature> pairFrames = tmpldrc.naiveMaterializeToRowFrames(rac);
+          Frame f = Frame.wrap(pairFrames.lhs);
+          try {
+            Iterables.getOnlyElement(outputChannels())
+                     .write(new FrameWithPartition(f, FrameWithPartition.NO_PARTITION));
+          }
+          catch (IOException e) {
+            throw new RuntimeException(e);
+          }
+          return Operator.Signal.GO;
```

Comment:
Instead of returning GO check if the frames can be paused. In such a case return that. Also need to test pausing frames through the MSQ framework correctly